### PR TITLE
Checked algebraic laws, refactoring

### PR DESCRIPTION
**Pull request by sphaso**
After a conversation with Tom, I wanted to check that Noether respects basic Functor and Monadic laws. The good news is that it does, the bad news is that `Either.bind` should ideally be more strict.      
For now I decided to go with a warning in the docs instead of a proper error.

After writing the tests, I decided to increase the inter-function dependency in `Maybe`. I doubt this will cause any issue in existing codebases, therefore I only consider this a "refactoring change" in the semver.

### CODE REVIEW

File: lib/noether/either.ex [hunk 1]
```diff
@@ -60,8 +60,7 @@ defmodule Noether.Either do
       {:error, "Value not found"}
   """
   @spec join(either()) :: either()
-  def join({:ok, {:ok, a}}), do: {:ok, a}
-  def join({:ok, {:error, a}}), do: {:error, a}
+  def join(a = {_, {_, _}}), do: bind(a, & &1)
```
Comment (sphaso) [inline lib/noether/either.ex:63]: In Haskell, `join` from `Control.Monad` is implemented in terms of `bind`.

File: lib/noether/maybe.ex [hunk 2]
```diff
@@ -9,10 +9,10 @@ defmodule Noether.Maybe do
 
   ## Examples
 
-      iex> map(nil, &Kernel.abs/1)
+      iex> Noether.Maybe.map(nil, &Kernel.abs/1)
```
Comment (sphaso) [inline lib/noether/maybe.ex:12]: The library that we use for Property Testing has a function called `map`, therefore I need to break the ambiguity.

File: mix.exs [hunk 3]
```diff
@@ -35,9 +35,9 @@ defmodule Noether.MixProject do
   # Run "mix help deps" to learn about dependencies.
   defp deps do
     [
-      {:credo, "~> 1.6.4", only: :dev, runtime: false},
+      {:credo, "~> 1.7", only: :dev, runtime: false},
```
Comment (sphaso) [inline mix.exs:38]: a long due version bump

File: lib/noether/maybe.ex [hunk 4]
```diff
@@ -85,12 +89,6 @@ defmodule Noether.Maybe do
...
   @spec choose(any(), fun1(), fun1()) :: any()
   def choose(a, f, g) when is_function(f, 1) and is_function(g, 1) do
-    b = f.(a)
-
-    if is_nil(b) do
-      g.(a)
-    else
-      b
-    end
+    maybe(a, f, map(a, g))
```
Comment (neslinesli93) [inline lib/noether/maybe.ex]: I don't know how this works in elixir, but are you sure that `map(a, g)` is not evaluated before the call to `maybe`? If that's the case and someone is doing side effects on `g` we are doing an involuntary breaking change... Maybe we can write a test where `g` (improperly) throws an exception and `f` returns a non nil value to check for the correct behavior

Comment (sphaso) [reply lib/noether/maybe.ex]: yes, I believe `map` will be evaluated before `maybe` since Elixir is not lazy. Very good point!

Comment (sphaso) [reply lib/noether/maybe.ex]: reverted

**Review by neslinesli93 — APPROVED** [review]

**Pull request merged**